lib/location/scan_wifi.c

```diff
@@ -0,0 +1,164 @@
+
+static struct net_if *wifi_iface;
+static struct net_mgmt_event_callback scan_wifi_net_mgmt_cb;
+
+static struct wifi_scan_result scan_results[CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT];
+static struct wifi_scan_info scan_wifi_info = {
+	.ap_info = scan_results,
+};
+static struct k_sem *scan_wifi_ready;
+
+struct wifi_scan_info *scan_wifi_results_get(void)
+{
+	if (scan_wifi_info.cnt  <= 1) {
+		if (scan_wifi_info.cnt == 1) {
+			/* Following statement seems to be true at least with HERE
+			 * (400: bad request).
+			 */
+			LOG_WRN("Retrieving a location based on a single Wi-Fi "
+				"access point is not possible, using only cellular data");
+		} else {
+			LOG_WRN("No Wi-Fi scanning results, using only cellular data");
+		}
+		return NULL;
+	}
+
+	return &scan_wifi_info;
+}
+
+int scan_wifi_start(struct k_sem *wifi_scan_ready)
+{
+	int ret;
+
+	scan_wifi_ready = wifi_scan_ready;
+
+	LOG_DBG("Triggering start of Wi-Fi scanning");
+
+	scan_wifi_info.cnt = 0;
+
+	__ASSERT_NO_MSG(wifi_iface != NULL);
+	ret = net_mgmt(NET_REQUEST_WIFI_SCAN, wifi_iface, NULL, 0);
+	if (ret) {
+		LOG_ERR("Failed to initiate Wi-Fi scanning: %d", ret);
+		ret = -EFAULT;
+		k_sem_give(wifi_scan_ready);
+		wifi_scan_ready = NULL;
+	}
+	return ret;
+}
+
+static void scan_wifi_result_handle(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_scan_result *entry = (const struct wifi_scan_result *)cb->info;
+	struct wifi_scan_result *current;
+
+	scan_wifi_info.cnt++;
+
+	if (scan_wifi_info.cnt <= CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT) {
+		current = &scan_wifi_info.ap_info[scan_wifi_info.cnt - 1];
+		*current = *entry;
+
+		LOG_DBG("scan result #%d stored: ssid %s, channel %d,"
+			" mac %02x:%02x:%02x:%02x:%02x:%02x",
+				scan_wifi_info.cnt,
+				current->ssid,
+				current->channel,
+				current->mac[0], current->mac[1], current->mac[2],
+				current->mac[3], current->mac[4], current->mac[5]);
+	} else {
+		LOG_WRN("Scanning result (mac %02x:%02x:%02x:%02x:%02x:%02x) "
+			"did not fit to result buffer - dropping it",
+				entry->mac[0], entry->mac[1], entry->mac[2],
+				entry->mac[3], entry->mac[4], entry->mac[5]);
+	}
+}
+
+static void scan_wifi_done_handle(struct net_mgmt_event_callback *cb)
+{
+	const struct wifi_status *status = (const struct wifi_status *)cb->info;
+
+	if (status->status) {
+		LOG_WRN("Wi-Fi scan request failed (%d).", status->status);
+	} else {
+		LOG_DBG("Scan request done.");
+	}
+
+	k_sem_give(scan_wifi_ready);
+	scan_wifi_ready = NULL;
+}
+
+void scan_wifi_net_mgmt_event_handler(
+	struct net_mgmt_event_callback *cb,
+	uint32_t mgmt_event,
+	struct net_if *iface)
+{
+	ARG_UNUSED(iface);
+
+	if (scan_wifi_ready != NULL) {
+		switch (mgmt_event) {
+		case NET_EVENT_WIFI_SCAN_RESULT:
+			scan_wifi_result_handle(cb);
+			break;
+		case NET_EVENT_WIFI_SCAN_DONE:
+			scan_wifi_done_handle(cb);
+			break;
+		default:
+			break;
+		}
+	}
+}
+
+int scan_wifi_cancel(void)
+{
+	k_sem_reset(scan_wifi_ready);
```

Comment:
Should this check if `scan_wifi_ready` is not `NULL`? What does `k_sem_reset()` do if called with a `NULL` pointer? There's a check in `scan_cellular.c`.

Reply:
Added NULL check